zcash_proofs/src/circuit/sapling.rs

```diff
@@ -70,22 +70,38 @@ pub struct Output<'a, E: JubjubEngine> {
 fn expose_value_commitment<E, CS>(
     mut cs: CS,
     value_commitment: Option<ValueCommitment<E>>,
-    params: &E::Params,
-) -> Result<Vec<boolean::Boolean>, SynthesisError>
-where
-    E: JubjubEngine,
-    CS: ConstraintSystem<E>,
+    params: &E::Params
+) -> Result<(Vec<boolean::Boolean>, Vec<boolean::Boolean>), SynthesisError>
+    where E: JubjubEngine,
+          CS: ConstraintSystem<E>
 {
+    // Witness the asset type
+    let asset_type = ecc::EdwardsPoint::witness(
+        cs.namespace(|| "asset_type"),
+        value_commitment.as_ref().map(|c| c.asset_type.value_commitment_generator(params)),
```

Comment:
The `value_commitment_generator` function is the group hash, which is what we want to be checking in the circuit, but is this code being run in the circuit? It seems like it isn't, unless somehow the `value_commitment_generator` function (which runs a loop) is being compiled to circuit constraints.

Reply:
Ok, so, here’s my newest question. I think you’re right that we have to add the value base to the note and then witness that the value commitment uses the value base and the value from the note, but I’m still not sure that we have to do the group hash in the note.

The purpose of including the value base would be to ensure, like you said, that you can’t change the value base where one value base is used in the Output circuit and a different value base is used in the Spend circuit for the same note (so note def needs to be changed)

The reason for my confusion was because I was thinking about ensuring consistency between two notes. I still don’t believe checking the value base is necessary for ensuring that two notes _within a transaction_ match because the value balance check should handle that.

But this doesn’t (necessarily) require that the value base is the result of a hash, though that would solve the slight malleability issue I described before

The other consideration is that doing the hash would allow the size of the note to stay smaller, potentially, than adding one G1 point to the note.  Is the size of the note a concern?